### PR TITLE
Bluebox weight proof.

### PR DESCRIPTION
**Pull request by fchirica**

### CODE REVIEW

File: src/full_node/full_node.py [hunk 1]
```diff
@@ -1363,3 +1376,297 @@ async def respond_transaction(
...
+                    new_finished_subslots[index] = new_subslot
+                    new_block = dataclasses.replace(block, finished_sub_slots=new_finished_subslots)
+                    break
+        if field_vdf == FieldVDF.CC_SP_VDF:
+            assert block.challenge_chain_sp_proof is not None
+            new_block = dataclasses.replace(block, challenge_chain_sp_proof=vdf_proof)
+        if field_vdf == FieldVDF.CC_IP_VDF:
+            new_block = dataclasses.replace(block, challenge_chain_ip_proof=vdf_proof)
+        assert new_block is not None
+        await self.block_store.add_full_block(new_block, block_record)
```
Comment (mariano54) [inline src/full_node/full_node.py:1487]: I think we should add a field in the DB that says: fully_compactified: True.
Then when doing the scan, we don't have to re-read 10,000,000 blocks from disk again. 
I think it's an absolute must.

**Review by mariano54 — COMMENTED** [review]
This looks really good. I have a few comments. The most important ones being:
- Don't re-read blocks from DB if they are compactified (very slow)
- Use header hash instead of height, and consider that there are multiple blocks per height
- Add some tests with a longer chain, in test_blockchain.py, if possible

File: src/full_node/full_node.py [hunk 2]
```diff
@@ -1363,3 +1376,297 @@ async def respond_transaction(
...
+                broadcast_list: List[timelord_protocol.NewProofOfTime] = []
+                min_height = 1
+                new_min_height = None
+                max_height = self.blockchain.get_peak_height()
+                if max_height is None:
+                    await asyncio.sleep(30)
+                    continue
+                batches_finished = 0
+                self.log.info("Scanning the blockchain for uncompact blocks.")
+                for h in range(min_height, max_height, 100):
```
[the remark this answers is not in the document]
Comment (fchirica) [reply src/full_node/full_node.py:1587]: The function get_header_blocks_in_range includes the end as well (e.g. we get blocks 100, 101, ..., 199, then the next start point starts at 200, as opposed to 199).

File: .github/workflows/build-test-macos-core.yml [hunk 3]
```diff
@@ -14,7 +14,7 @@ jobs:
   build:
     name: MacOS Core Tests
     runs-on: ${{ matrix.os }}
-    timeout-minutes: 70
+    timeout-minutes: 210
```
[the remark this answers is not in the document]
Comment (fchirica) [reply .github/workflows/build-test-macos-core.yml:17]: test_simulation is super slow, I set it temporarily to see if we can pass with big timeout.

File: tests/core/test_simulation.py [hunk 4]
```diff
@@ -22,37 +23,59 @@
...
-        node1, node2, _, _, _, _, _, server1 = simulation
+    async def test_simulation_1(self, simulation, extra_node):
+        node1, node2, _, _, _, _, _, _, _, server1 = simulation
         await server1.start_client(PeerInfo(self_hostname, uint16(21238)))
         # Use node2 to test node communication, since only node1 extends the chain.
         await time_out_assert(1000, node_height_at_least, True, node2, 7)
 
+        async def has_compact(node1, node2):
+            peak_height_1 = node1.full_node.blockchain.get_peak_height()
+            headers_1 = await node1.full_node.blockchain.get_header_blocks_in_range(1, peak_height_1)
```
[the remark this answers is not in the document]
Comment (fchirica) [reply tests/core/test_simulation.py:46]: we need just one compact of each type, not every one to be compact until peak height. going till the end increases the chance to find one, given the randomness of the whole process.

File: src/full_node/block_store.py [hunk 5]
```diff
@@ -319,3 +320,23 @@ async def set_peak(self, header_hash: bytes32) -> None:
...
+        )
+        row = await cursor.fetchone()
+        await cursor.close()
+        if row is None:
+            return None
+        return bool(row[0])
+
+    async def get_first_not_compactified(self, min_height: int) -> Optional[int]:
+        cursor = await self.db.execute(
+            "SELECT MIN(height) from full_blocks WHERE is_fully_compactified=0 AND height>=?", (min_height,)
```
Comment (mariano54) [inline src/full_node/block_store.py:336]: Can you create a DB index to speed this up?

File: tests/core/fixtures.py [hunk 6]
```diff
@@ -63,7 +63,18 @@ async def default_20000_blocks():
     return persistent_blocks(20000, f"test_blocks_20000_{block_format_version}.db")
 
 
-def persistent_blocks(num_of_blocks: int, db_name: str, seed: bytes = b"", empty_sub_slots=0):
+@pytest.fixture(scope="session")
+async def default_10000_blocks_compact():
```
Comment (mariano54) [inline tests/core/fixtures.py:67]: Did you add these blocks to the block cache repo? That will speed up tests.

**Review by mariano54 — APPROVED** [review]
Looks good. The things remaining are:
* Getting weight proof creation in
* Potentially splitting out the simulation tests if they are too slow. Is this what is causing the slowdown? We don't want to make CI much slower

**Pull request merged**